tokio-uds/tests/datagram.rs

```diff
@@ -0,0 +1,85 @@
+#![cfg(unix)]
+
+extern crate bytes;
+extern crate futures;
+extern crate tempdir;
+extern crate tokio;
+extern crate tokio_codec;
+extern crate tokio_uds;
+
+use tokio_uds::*;
+
+use std::str;
+
+use bytes::BytesMut;
+
+use tokio::io;
+use tokio::runtime::current_thread::Runtime;
+
+use tokio_codec::{Decoder, Encoder};
+
+use futures::{Future, Sink, Stream};
+
+use tempdir::TempDir;
+
+struct StringDatagramCodec;
+
+/// A codec to decode datagrams from a unix domain socket as utf-8 text messages.
+impl Encoder for StringDatagramCodec {
+    type Item = String;
+    type Error = io::Error;
+
+    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        dst.extend_from_slice(&item.into_bytes());
+        Ok(())
+    }
+}
+
+/// A codec to decode datagrams from a unix domain socket as utf-8 text messages.
+impl Decoder for StringDatagramCodec {
+    type Item = String;
+    type Error = io::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        let decoded = str::from_utf8(buf)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?
+            .to_string();
+
+        Ok(Some(decoded))
+    }
+}
+
+#[test]
+fn framed_echo() {
+    let dir = TempDir::new("tokio-uds-tests").unwrap();
+    let server_path = dir.path().join("server.sock");
+    let client_path = dir.path().join("client.sock");
+
+    let mut rt = Runtime::new().unwrap();
+
+    {
+        let socket = UnixDatagram::bind(&server_path).unwrap();
+        let server = UnixDatagramFramed::new(socket, StringDatagramCodec);
+
+        let (sink, stream) = server.split();
+
+        let echo_stream = stream.map(|(msg, addr)| (msg, addr.as_pathname().unwrap().into()));
+
+        // spawn echo server
+        rt.spawn(echo_stream.forward(sink)
+                 .map_err(|e| panic!("err={:?}", e))
+                 .map(|_| ()));
+    }
+
+    {
+        let socket = UnixDatagram::bind(&client_path).unwrap();
```

Comment:
Is the client suppose to `bind()` or `connect()`?

Reply:
Oh I missed that the path was different. I thought you were connecting to the server's path.

Reply:
Talked this out more here: https://gitter.im/tokio-rs/tokio?at=5b329778d2abe466888c8f1b

Reply:
I got confused here between `SOCK_STREAM` and `SOCK_DGRAM`